### ISSUE TITLE
CloudSuite incompatibility with Kubernetes

### ISSUE THREAD

**Issue by Remit** [parsa-epfl/cloudsuite#226]
Hello,

I was considering to use CloudSuite to benchmark some features that I've added to Kubernetes, but I've found out that Cloudsuite requires a substantial configuration in data-analytics benchmark to establish overlay networking between master and slave.

Prerequisites:
- single-node Kubernetes cluster with enough capacity, untainted node, and Calico as network plugin
- master-deployment.yaml (deploys and runs successfully):

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: cloudsuite-data-analytics-master-deployment
  labels:
    app: cloudsuite-data-analytics-master
spec:
  replicas: 1
  selector:
    matchLabels:
      app: cloudsuite-data-analytics-master
  template:
    metadata:
      name: master
      labels:
        app: cloudsuite-data-analytics-master

    spec:
      hostname: master
      tolerations:
      - key: "allowed"
        operator: "Equal"
        # The value should be corrected for the real run. Also, one can use env variables.
        value: "k8instance"
        effect: "NoSchedule"
      containers:
      - name: master
        image: cloudsuite/data-analytics:latest
        # Reference for command arguments for entrypoint: https://kubernetes.io/docs/tasks/inject-data-application/define-command-argument-container/
        args: ["master"]
        ports:
          # Reference for Hadoop ports: https://ambari.apache.org/1.2.3/installing-hadoop-using-ambari/content/reference_chap2_1.html
          - name: http
            containerPort: 50070
          - name: https
            containerPort: 50470
          - name: ipc1
            containerPort: 8020
          - name: ipc2
            containerPort: 9000
        resources:
          requests:
            cpu: "0.5"
            memory: "4Gi"
          limits:
            cpu: "0.5"
            memory: "4Gi"
```

- worker-deployment.yaml (deploys and runs successfully):
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: cloudsuite-data-analytics-worker-deployment
  labels:
    app: cloudsuite-data-analytics-worker
spec:
  replicas: 1
  selector:
    matchLabels:
      app: cloudsuite-data-analytics-worker
  template:
    metadata:
      name: slave01
      labels:
        app: cloudsuite-data-analytics-worker
        numapolicy: numaaware
        # For baseline test only:
        testpolicy: antinuma

    spec:
      hostname: slave01
      tolerations:
      - key: "allowed"
        operator: "Equal"
        # The value should be corrected for the real run. Also, one can use env variables.
        value: "k8instance"
        effect: "NoSchedule"

      containers:
      - name: slave01
        image: cloudsuite/hadoop:latest
        # Reference for command arguments for entrypoint: https://kubernetes.io/docs/tasks/inject-data-application/define-command-argument-container/
        args: ["slave"]
        ports:
          # Reference for Hadoop ports: https://ambari.apache.org/1.2.3/installing-hadoop-using-ambari/content/reference_chap2_1.html
          - name: http
            containerPort: 50075
          - name: https
            containerPort: 50475
          - name: data
            containerPort: 50010
          - name: ipc
            containerPort: 50020
        resources:
          requests:
            cpu: "1.0"
            memory: "4Gi"
          limits:
            cpu: "1.0"
            memory: "4Gi"
```

Situation:
When running `benchmark` in master pod, the following error happens:

> 19/08/07 13:37:31 WARN hdfs.DFSClient: DataStreamer Exception
> org.apache.hadoop.ipc.RemoteException(java.io.IOException): File /user/root/wiki._COPYING_ could only be replicated to 0 nodes instead of minReplication (=1).  There are 0 datanode(s) running and no node(s) are excluded in this operation.
>         at org.apache.hadoop.hdfs.server.blockmanagement.BlockManager.chooseTarget4NewBlock(BlockManager.java:1628)
>         at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.getNewBlockTargets(FSNamesystem.java:3121)
>         at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.getAdditionalBlock(FSNamesystem.java:3045)
>         at org.apache.hadoop.hdfs.server.namenode.NameNodeRpcServer.addBlock(NameNodeRpcServer.java:725)
>         at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolServerSideTranslatorPB.addBlock(ClientNamenodeProtocolServerSideTranslatorPB.java:493)
>         at org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos$ClientNamenodeProtocol$2.callBlockingMethod(ClientNamenodeProtocolProtos.java)
>         at org.apache.hadoop.ipc.ProtobufRpcEngine$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine.java:616)
>         at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:982)
>         at org.apache.hadoop.ipc.Server$Handler$1.run(Server.java:2217)
>         at org.apache.hadoop.ipc.Server$Handler$1.run(Server.java:2213)
>         at java.security.AccessController.doPrivileged(Native Method)
>         at javax.security.auth.Subject.doAs(Subject.java:422)
>         at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1746)
>         at org.apache.hadoop.ipc.Server$Handler.run(Server.java:2213)
> 
>         at org.apache.hadoop.ipc.Client.call(Client.java:1476)
>         at org.apache.hadoop.ipc.Client.call(Client.java:1413)
>         at org.apache.hadoop.ipc.ProtobufRpcEngine$Invoker.invoke(ProtobufRpcEngine.java:229)
>         at com.sun.proxy.$Proxy10.addBlock(Unknown Source)
>         at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolTranslatorPB.addBlock(ClientNamenodeProtocolTranslatorPB.java:418)
>         at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
>         at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
>         at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
>         at java.lang.reflect.Method.invoke(Method.java:498)
>         at org.apache.hadoop.io.retry.RetryInvocationHandler.invokeMethod(RetryInvocationHandler.java:191)
>         at org.apache.hadoop.io.retry.RetryInvocationHandler.invoke(RetryInvocationHandler.java:102)
>         at com.sun.proxy.$Proxy11.addBlock(Unknown Source)
>         at org.apache.hadoop.hdfs.DFSOutputStream$DataStreamer.locateFollowingBlock(DFSOutputStream.java:1588)
>         at org.apache.hadoop.hdfs.DFSOutputStream$DataStreamer.nextBlockOutputStream(DFSOutputStream.java:1373)
>         at org.apache.hadoop.hdfs.DFSOutputStream$DataStreamer.run(DFSOutputStream.java:554)
> put: File /user/root/wiki._COPYING_ could only be replicated to 0 nodes instead of minReplication (=1).  There are 0 datanode(s) running and no node(s) are excluded in this operation.

When logging inside each container, I was able to see the other one by their IPs, e.g. 192.168.180.8 or 192.168.180.9. I also tried to change /etc/hosts and manually execute entrypoint.sh, but still no success in connection, e.g. slave tells:

> 2019-08-07 13:13:05,662 WARN org.apache.hadoop.hdfs.server.datanode.DataNode: Problem connecting to server: master:8020

So, to sum up: both master and slave work well independently (e.g. web-UI is in place) + they can see each other with ssh, but connection between them is not established. I think that there could be some networking issue, but I don't on which side it is: Kubernetes? settings of the Hadoop?

I have no experience with Hadoop, so I was wondering if you could help me figure out the solution.
Also, as a suggestion - maybe it would make sense to make Kubernetes deployments for the provided benchmarks? It would really expand the use of the Suite, I believe.

--
Best regards,
Vladimir

**Comment by ZiluTian** [comment]
Hello,
 
Can you please try using images zilutian/hadoop:latest and zilutian/data-analytics:latest and confirm if the problem is still relevant? 

Thanks, 
Zilu

**Comment by Remit** [comment]
Hi @ZiluTian ,

Thanks for the prompt reply.
Still, the same problem remains:

> 19/08/07 14:33:10 WARN hdfs.DataStreamer: DataStreamer Exception
> org.apache.hadoop.ipc.RemoteException(java.io.IOException): File /user/root/wiki._COPYING_ could only be replicated to 0 nodes instead of minReplication (=1).  There are 0 datanode(s) running and no node(s) are excluded in this operation.
>         at org.apache.hadoop.hdfs.server.blockmanagement.BlockManager.chooseTarget4NewBlock(BlockManager.java:1814)
>         at org.apache.hadoop.hdfs.server.namenode.FSDirWriteFileOp.chooseTargetForNewBlock(FSDirWriteFileOp.java:265)
>         at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.getAdditionalBlock(FSNamesystem.java:2563)
>         at org.apache.hadoop.hdfs.server.namenode.NameNodeRpcServer.addBlock(NameNodeRpcServer.java:846)
>         at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolServerSideTranslatorPB.addBlock(ClientNamenodeProtocolServerSideTranslatorPB.java:510)
>         at org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos$ClientNamenodeProtocol$2.callBlockingMethod(ClientNamenodeProtocolProtos.java)
>         at org.apache.hadoop.ipc.ProtobufRpcEngine$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine.java:503)
>         at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:989)
>         at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:871)
>         at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:817)
>         at java.security.AccessController.doPrivileged(Native Method)
>         at javax.security.auth.Subject.doAs(Subject.java:422)
>         at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1889)
>         at org.apache.hadoop.ipc.Server$Handler.run(Server.java:2606)
> 
>         at org.apache.hadoop.ipc.Client.getRpcResponse(Client.java:1493)
>         at org.apache.hadoop.ipc.Client.call(Client.java:1439)
>         at org.apache.hadoop.ipc.Client.call(Client.java:1349)
>         at org.apache.hadoop.ipc.ProtobufRpcEngine$Invoker.invoke(ProtobufRpcEngine.java:227)
>         at org.apache.hadoop.ipc.ProtobufRpcEngine$Invoker.invoke(ProtobufRpcEngine.java:116)
>         at com.sun.proxy.$Proxy10.addBlock(Unknown Source)
>         at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolTranslatorPB.addBlock(ClientNamenodeProtocolTranslatorPB.java:444)
>         at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
>         at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
>         at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
>         at java.lang.reflect.Method.invoke(Method.java:498)
>         at org.apache.hadoop.io.retry.RetryInvocationHandler.invokeMethod(RetryInvocationHandler.java:422)
>         at org.apache.hadoop.io.retry.RetryInvocationHandler$Call.invokeMethod(RetryInvocationHandler.java:165)
>         at org.apache.hadoop.io.retry.RetryInvocationHandler$Call.invoke(RetryInvocationHandler.java:157)
>         at org.apache.hadoop.io.retry.RetryInvocationHandler$Call.invokeOnce(RetryInvocationHandler.java:95)
>         at org.apache.hadoop.io.retry.RetryInvocationHandler.invoke(RetryInvocationHandler.java:359)
>         at com.sun.proxy.$Proxy11.addBlock(Unknown Source)
>         at org.apache.hadoop.hdfs.DataStreamer.locateFollowingBlock(DataStreamer.java:1845)
>         at org.apache.hadoop.hdfs.DataStreamer.nextBlockOutputStream(DataStreamer.java:1645)
>         at org.apache.hadoop.hdfs.DataStreamer.run(DataStreamer.java:710)
> put: File /user/root/wiki._COPYING_ could only be replicated to 0 nodes instead of minReplication (=1).  There are 0 datanode(s) running and no node(s) are excluded in this operation.
> Mahout: seqwiki
> MAHOUT_LOCAL is not set; adding HADOOP_CONF_DIR to classpath.

**Comment by ZiluTian** [comment]
Hi, 
Thanks for the confirmation. So I tested with Docker Swarm using Overlay network driver and the benchmark runs fine. It appears to be a K8-specific config problem. I found this webpage describing how to run Hadoop with Kubernetes, and it seems like additional environmental variables are required to be set for K8 to work: https://blog.hasura.io/getting-started-with-hdfs-on-kubernetes-a75325d4178c/ Can you please give it a try? Unfortunately I am not very familiar with Kubernetes. I will try to reproduce the problem and get back to you in the next few days if you are still unable to get it to work. 

Regards, 
Zilu

**Comment by Remit** [comment]
@ZiluTian thanks for the suggestion.
I've tried to elaborate on services as author suggests, it did help to a certain extent. Now, both containers are able to see each other by their names. But... That works only if I use curl. For instance, when I run in from worker, I get:

> It looks like you are making an HTTP request to a Hadoop IPC port. This is not the correct port for the web interface on this daemon.

That's ok. But when I start the entrypoint.sh script with slave as argument, I get:

> 2019-08-07 20:27:01,809 WARN org.apache.hadoop.hdfs.server.datanode.DataNode: Problem connecting to server: master:8020

Just to be sure, here is the /etc/hosts as created by Kubernetes:

> # Kubernetes-managed hosts file.
> 127.0.0.1       localhost
> ::1     localhost ip6-localhost ip6-loopback
> fe00::0 ip6-localnet
> fe00::0 ip6-mcastprefix
> fe00::1 ip6-allnodes
> fe00::2 ip6-allrouters
> 192.168.180.6   slave01

Any ideas?

**Comment by ZiluTian** [comment]
Hello, 

Can you please check the logs for master and worker and confirm the following things? In docker you can use command ```docker logs *container_name*```, there should be similar commands for Kubernetes. Or you can simply copy the logs folder to your host: 
```docker cp master:/opt/hadoop-2.9.1/logs master_log```
```docker cp slave01:/opt/hadoop-2.9.1/logs slave_log```

Master: 
IPC Server has started properly 
```IPC Server listener on 8020: starting``` 
NameNode RPC server is up 
```NameNode RPC up at: master/*ip*:8020```

Slave: 
datanode has started properly 
```
starting datanode, logging to ... 
starting nodemanager, logging to ...
```

Regards, 
Zilu

**Comment by Remit** [comment]
Hey @ZiluTian,

I've attached the logs:

[hadoop--namenode-master.log](https://github.com/parsa-epfl/cloudsuite/files/3481257/hadoop--namenode-master.log)
[mapred--historyserver-master.log](https://github.com/parsa-epfl/cloudsuite/files/3481258/mapred--historyserver-master.log)
[yarn--resourcemanager-master.log](https://github.com/parsa-epfl/cloudsuite/files/3481259/yarn--resourcemanager-master.log)
[hadoop--datanode-master.log](https://github.com/parsa-epfl/cloudsuite/files/3481262/hadoop--datanode-master.log)
[yarn--nodemanager-master.log](https://github.com/parsa-epfl/cloudsuite/files/3481263/yarn--nodemanager-master.log)


Also, the configuration of some files from datanode as suggested in the log:
- hdfs-site.xml

```
<?xml version="1.0"?>
<?xml-stylesheet type="text/xsl" href="configuration.xsl"?>
<configuration>
  <property>
    <name>dfs.replication</name>
    <value>1</value>
  </property>
  <property>
    <name>dfs.namenode.name.dir</name>
    <value>file:/root/hdfs/namenode</value>
  </property>
  <property>
    <name>dfs.datanode.data.dir</name>
    <value>file:/root/hdfs/datanode</value>
  </property>
</configuration>
```
- core-site.xml

```
<?xml version="1.0"?>
<?xml-stylesheet type="text/xsl" href="configuration.xsl"?>
<configuration>
  <property>
    <name>fs.defaultFS</name>
    <value>hdfs://master:8020</value>
  </property>
</configuration>

```

**Comment by ZiluTian** [comment]
Hello, 
How did you deploy the master and the worker? Can you list the exact commands you used to launch the cluster? 

Regards, 
Zilu

**Comment by ZiluTian** [comment]
Hello, 

I think the problem might be that you specify both master and worker as "kind: deployment", when they really should be pods within the same deployment. Please try it and let me know if it helps. 

Regards, 
Zilu

**Comment by Remit** [comment]
@ZiluTian thanks for the suggestion - I've tried to create two raw pods (as deployment does not support different types of pods in Kubernetes), but it did not work. Still the same problem + now I cannot reach one pod from the other via curl anymore (it was previously ensured by the combination deployment+service).

Maybe I need to make some additional settings in HDFS's /etc?

Best regards,
Vladimir

**Comment by ZiluTian** [comment]
Hello, 
It is quite unlikely. The only assumption of the HDFS config is that datanode (slave) can reach namenode (master) at the name "master". My speculation is that the problem is related to Kubernetes' network config, either the slave doesn't share the same network with master somehow which resulted in master being unreachable, or the two can see each other but under different names. 

Regards, 
Zilu

**Comment by neo-apz** [comment]
I'm closing this for now.
Feel free to reopen it if you have any further questions.

**Comment by cnauroth** [comment]
This issue is closed, but I wanted to mention that another possibility is that you were running into this Hadoop bug:

[HADOOP-15129](https://issues.apache.org/jira/browse/HADOOP-15129): Datanode caches namenode DNS lookup failure and cannot startup (GitHub PR [#3348](https://github.com/apache/hadoop/pull/3348))

If the DataNode starts while the NameNode host isn't even resolvable in DNS, then the DataNode does not attempt to re-resolve.  I've made available a patch that moves the DNS resolution into a connection retry loop.

This patch is not yet accepted or included in an Apache Hadoop release.
